### PR TITLE
FOGL-8123 Add performance monitors to north service

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

**Review by pintomax — DISMISSED** [review]

File: C/services/north/north.cpp [hunk 1]
```diff
@@ -357,6 +360,7 @@ void NorthService::start(string& coreAddress, unsigned short corePort)
 		northConfig.setDescription("North");
 		m_mgtClient->addCategory(northConfig, true);
 
+		// Fetch Confguration
```
Comment (RVerhoeff) [inline C/services/north/north.cpp:363]: Typo: Confguration -> Configuration

**Review by RVerhoeff — DISMISSED** [review]

**Review by pintomax — APPROVED** [review]

**Review by pintomax — APPROVED** [review]

File: docs/tuning_fledge.rst [hunk 2]
```diff
@@ -81,6 +81,61 @@ The south services within Fledge each have a set of advanced configuration optio
...
+      - The rate that data is being ingested at as a result of flow control throttling.
+      - This counter is more for information as to what might make a reasonable ingest rate the system can sustain with the current configuration. It is useful as it gives a good idea of how far away from your desired performance the current configuration of the system is currently
+    * - storedReadings
+      - The readings successfully sent to the storage layer.
+      - This counter gives an indication of the bandwidth available from the service to the storage engine. This should be at least as high as the ingest rate if data is not to accumulate in buffers within the storage. Altering the maximum latency and maximum buffered readings advanced settings in the south server can impact this throughput.
+    * - resendQueued
+      - The number of readings queued for resend. Note that readings may be queued for resend multiple times if the resend also failed.
+      - This is a good indication of overload conditions within the storage engine. Consistent high values of this counter point to the need to improve the performance of the storage layer.
+    * - removedReadings
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:134]: I tried these but not able to see these counters into monitors table. Are they working?

File: docs/tuning_fledge.rst [hunk 3]
```diff
@@ -126,6 +181,56 @@ In a similar way to the south services, north services and tasks also have advan
...
+
+  - The minimum value of the counter observed within the current minute
+
+  - The maximum value of the counter observed within the current minute
+
+  - The average value of the counter observed within the current minute
+
+  - The number of samples of the counter collected within the current minute
+
+In the current release the performance counters can only be retrieved by director access to the configuration and statistics database, they are stored in the *monitors* table. Future releases will include tools for the retrieval and analysis of these performance counters.
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:201]: This is a repetitive stuff as same as in section of `South Service Advanced Configuration-> Performance Counters`. We may add this it as  label with reference link. (Likewise Summary of Performance Counter)

File: docs/tuning_fledge.rst [hunk 4]
```diff
@@ -126,6 +181,56 @@ In a similar way to the south services, north services and tasks also have advan
...
+      - This is a counter of the number of blocks that are waiting to be sent in the north service
+      - if this figure is more than a couple of blocks it is an indication that the north plugin is failing to sent complete blocks of data and that partial blocks are failing. Reducing the block size may improve the situation and reduce the amount of storage required in the north service.
+    * - Total readings buffered
+      - This is a count of the total number of readings buffered within the north service.
+      - This should be equivalent to 2 or 3 blocks size worth of readings. If it is high then it is an indication that the north plugin is not able to sustain a high enough data rate to match the ingest rates of the system.
+    * - Readings sent
+      - This gives an indication, for each block, how many readings are sent in the block.
+      - This should typically match the blocks read, if not it is an indication of failures to send data by the north plugin.
+    * - Percentage readings sent
+      - Closely related to the above the s the percentage of each block read that was actually sent.
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:228]: `the s the` ?

File: C/services/north/north.cpp [hunk 5]
```diff
@@ -402,7 +406,18 @@ void NorthService::start(string& coreAddress, unsigned short corePort)
 
 		m_storage->registerManagement(m_mgtClient);
 
-		// Fetch Confguration
+		// Setup the performance monitor
+		m_perfMonitor = new PerformanceMonitor(m_name, m_storage);
+
+		if (m_configAdvanced.itemExists("perfmon"))
```
Comment (ashish-jabble) [inline C/services/north/north.cpp:412]: I believe consistency wise this should be in camelCase i.e `perfMon` as other advanced config items were with same naming convention.

Same applies to South side

File: docs/tuning_fledge.rst [hunk 6]
```diff
@@ -81,6 +81,61 @@ The south services within Fledge each have a set of advanced configuration optio
 
    The *Statistics Collection* setting will not remove any existing statistics, these will remain and remain to be represented in the statistics history. This only impacts new values that are collected. It is recommended that this be set before a service is started for the first time if the desire it to have no statistics values recorded for either assets or the service.
 
+
+  - *Performance Counters* - This option allows for the collection of performance counters that can be used to help tune the south service.
+
+Performance Counters
+--------------------
+
+A number of performance counters can be collected in the south service to help characterise the performance of the service. This is intended to provide input into the tuning of the service and the collection of these counters should not be left on during production use of the service.
```
Comment (ashish-jabble) [inline docs/tuning_fledge.rst:90]: > these counters should not be left on during production 

@MarkRiddoch Do you mean here
a) Its a recommendation/notice to users; Don't use this property in production environment?
May be we can **add as NOTE** here with highlight

May be we can enable/disable this config item on the basis of developer option settings in GUI? That also means we need another optional attribute property required to handle this, as _readonly_ config item is not fit here.

File: docs/tuning_fledge.rst (hunk 2 above)
Comment (MarkRiddoch) [reply docs/tuning_fledge.rst:134]: These are implemented inn a different pull request but have been documented here.

File: docs/tuning_fledge.rst (hunk 3 above)
Comment (MarkRiddoch) [reply docs/tuning_fledge.rst:201]: It was doen deliberately as soem users may just jump to the section that is of interest to them.

**Pull request merged**